Nordigen Transaction history UI

dshoreman opened the pull request:
## tl;dr

Adds a new screen to the Nordigen flow that enables selecting a custom transaction history.

---


![nordigen-history-select](https://github.com/user-attachments/assets/20c4cd88-5ddb-4a8c-8267-e70b59314217)

> ***NOTE: Pending rebase!** I started this before #10410, so Git hates me.*
> 
> I'm *fairly* sure this PR is still relevant (unless it's been done already).
> There's some overlap in the EUA code between that PR and this one. Rebase could take (tomorrow me) a little thinking so I wanted to post as a draft to get initial thoughts, make sure I'm heading in the right direction etc


### Key Points

* Enabled country list
  * Wasn't possible before due to how event listeners were done in the example JS
  * Achieved by wrapping it up in a `MutationObserver`
* Fixed 500 on the first `token-invalid` return
  * `$request->getCompany()` calls `getTokenContent()` internally
  * Wait 'til cache expires (1hr), token/context becomes `null`
  * The `if (!$context)` check happened *after* `getCompany()`, triggering the 500
* New page in Nordigen Connect UI for selecting number of days' history
  * Options autogenerated up to institution maximum
  * Multiples of 30
    * Skips last option if it's close (e.g. if max is 250, don't show 240)
    * 360 becomes 365 (if max > 13mo, otherwise see above)
  * Auto-sets transaction history from agreement on renewal
    * EUA ID extracted from "account expired" error


### Other Stuff

More details in commit messages, but a *lot* of the main diff is general cleanup stuff. When it's ready you *miiight* want to check the individual commit diffs first. If necessary I can move cleanup stuff to its own PR (or drop some of it entirely).

dshoreman committed:
Nordigen: Pass custom agreement with max history

In order to change the total days of transactions we can request, a
custom end user agreement is needed. There *are* methods to find
existing agreements, but assuming nothing else breaks it's probably
fairly safe to just go ahead and create one.

If it works, it only runs once... as far as I can tell.

dshoreman committed:
Nordigen: Use institution max for bank integration

While we set the days in the agreement, how many we fetch when
processing transactions depends on the `from_date` of the bank
integration. We could call `getInstitution($id)` on the API, but
consistency would dictate adding a wrapper in the Nordigen class and
it's already being called within `$nordigen->getAccount()` so it makes
more sense to extract the data in the account transformer instead.

App\Jobs\Bank\ProcessBankTransactionsNordigen also sets a from date, but
that's only used when not set on the bank integration so it can be left
at the Nordigen default of 90 days worth of transaction history.

dshoreman committed:
Enable country filtering for Nordigen bank list

Enabling the country filter stops the selector loop from working because
instead of a `.ob-list-institution` it starts with `.ob-country-list`
items which then get replaced on click so banks don't exist on load.

By wrapping the institution loop in an observer it'll automatically
attach the events when a country is selected and it works wonderfully.

dshoreman committed:
Allow setting number of days of transactions

If e.g. `&tx_days=365` is added to the URL, this will use it to set the
amount of history that's fetched from Nordigen on sync, with basic
safeguards to prevent it being set too low or above the bank's max.

dshoreman committed:
Add selection screen for transaction history

Instead of directing the user straight to GoCardless, when a bank is
selected this will replace the institution list with a list of options
from 30 days—up to the bank's transaction limit—in 30 day increments.

dshoreman committed:
Improve Nordigen error-handling code

All failures boil down to about 3 variations - one with reason/lang, one
that has context defined but no company data, and a full variant that
also fills the company and account keys in the view.

The only other difference is redirect url coming from config when it's
not available in the context, so to simplify things this commit merges
all the failure view responses into one unified `failed` method.

dshoreman committed:
Fine-tune transaction history options

Replaces 360 days with 365, skips the last auto-generated option to
avoid getting e.g. 720 and 730 as options when it's close to max, and
gradually increases the accumulator so options aren't so overwhelming.

Since the loop will no longer generate the final option (0 < 15), this
commit also removes the `% 30` check so the real max is always used.

dshoreman committed:
Fix back/close button display in Nordigen modal

When the country filtering mode is enabled, it adds a "Go back" button
but the close button is right on top of it. This moves the close button
to the right to avoid any click conflicts.

To fix an annoying display bug with the top offset, an observer is used
to override the default `display: none` of the back button to keep its
flex mode but with a hidden visibility so the UI doesn't move around.

dshoreman committed:
Find old EUA to get tx_days for new agreement

The end user agreement ID isn't stored with the bank integration, but it
*is* returned as part of the error for an expired account which works
perfectly for the case of renewing an expired requisition.

When `tx_days` isn't set in the request (i.e. it's a renewal) we instead
extract the EUA ID from the account error after getting the integration,
then once we have the EUA we're able to restore the old tx_days setting.

Since the BankIntegration query is used in both endpoints, this moves it
to a method with `firstOrFail()` rather than `first()` which also allows
for a cleanup of the integration saving code with try/catch/finally to
make it a little clearer which values apply to both new+existing cases.

dshoreman committed:
Fix error on token-invalid failure when expired

The cache data is set to expire after an hour, at which point the call
to `$request->getCompany()` will fail because it depends internally on
the `company_key` in the response of `getTokenContent()` which is null.

By moving the company *after* the context check, we avoid the exception
but need to manually set lang as there's no context for company data.

dshoreman committed:
Cleanup, add return types, normalise quotes

paulwer commented:
@dshoreman hi,
I also wanted to set max historical days and max validity based on institutions endpoint for the agreement and the initial import next.

You can reach out via slack to me, so we can discuss the way to progress if you like. Name: Paul Werner

Very good cleanups btw, like them :)

paulwer commented:
@turbo124 you can add me as an reviewed if you like :)

paulwer commented on app/Http/Controllers/Bank/NordigenController.php:
```diff
@@ -17,270 +17,254 @@
                 $bank_integration->nordigen_account_id = $nordigen_account['id'];
                 $bank_integration->bank_account_type = $nordigen_account['account_type'];
                 $bank_integration->bank_account_name = $nordigen_account['account_name'];
-                $bank_integration->bank_account_status = $nordigen_account['account_status'];
```
please explain this change :)

dshoreman replied:
From f37edfdabaf501c3cbb3fbfaa77350d93a5954ae - Instead of an if/else block where some things (like status and balance) are duplicated, the BankIntegration query now throws an exception so the non-unique bits are applied once in the `finally` instead.

paulwer commented on app/Http/Controllers/Bank/NordigenController.php:
```diff
@@ -17,270 +17,254 @@
                 $bank_integration->nordigen_institution_id = $nordigen_account['provider_id'];
                 $bank_integration->provider_name = $nordigen_account['provider_name'];
                 $bank_integration->nickname = $nordigen_account['nickname'];
-                $bank_integration->balance = $nordigen_account['current_balance'];
```
please explain this change :)

paulwer commented on app/Http/Controllers/Bank/NordigenController.php:
```diff
@@ -17,270 +17,254 @@
+
+        // Renewals have an Institution ID, but bypass the history selection screen
+        // and thus lack the history setting, so we can find the Agreement ID here.
+        if (!isset($data['tx_days'])) {
+            // Query the integration so we get the correct Account ID
+            $integration = $this->findIntegrationBy('institution', $institution, $company);
+
+            // Extract the EUA ID from the expired account error
+            $match = '/End User Agreement \(EUA\) ([0-9a-f-]+) has expired/';
+            $nordigenAccount = $nordigen->getAccount($integration->nordigen_account_id);
+            $euaId = preg_replace($match, '${1}', $nordigenAccount['error']);
+
+            // Fetch the old agreement and maintain its history setting
+            $data['tx_days'] = $nordigen->getAgreement($euaId)['max_historical_days'];
+        }
```
please explain the expected behavior here pls for me. I think this could have some difficulties, if a person has multiple accounts at one institutionId, then we can have same problems here.

a user can have multiple accounts in 2 different ways: 
1. multiple accounts per request (with the same credentials) f.ex. multiple bank_accounts within one account
2. multiple bank_accounts from different accounts (credentials) 

i think you should considder rework this section to use the defaults eigher from existing accounts or by the defaults on the institution. maybe only from the institution to reduce complexity

dshoreman replied:
You're right, it's definitely not *perfect*.

The red triangle on an expired integration is still treated as a completely new connection, except the Institution ID is pre-set. The */nordigen/connect* request knows nothing about the integration, so it's a best-effort situation with the information available.

I'd read previously that only one integration/account was officially supported because of how transactions were handled/linked or something, I'm not sure... @turbo124 can you give some insight?

---

#### Options
* Since #10404, upstream disabled state could be used to filter integrations
  * More reliable, but still uses some guesswork.
* If the BankIntegration ID was in the request we could directly query it
  * Injecting the ID into token could work?
  * Separate route (e.g. `/nordigen/{bank_integration}/reconnect/{token}`) would be best
    * ...but requires rewriting half the controller
  * Injecting the ID into token could work?

**Question:** When you're on the integration list and click the expired button, where does the `institution_id` get added into the token/cache?

paulwer replied:
bank_integration_id should be a unique id from nordigen, which is the same for each account, therefore integrations with disabled upsteam are reactivated on reconnect. (not sure if also archieved ones)
Also we have descided to dont implement 2 routes, as of to reduce complexity. after a callback (confirm) the right account is upserted.

paulwer commented on resources/views/bank/nordigen/handler.blade.php:
```diff
@@ -27,7 +30,7 @@
         logoUrl: "{{ ($account ?? false) && !$account->isPaid() ? asset('images/invoiceninja-black-logo-2.png') : (isset($company) && !is_null($company) ? $company->present()->logo() : asset('images/invoiceninja-black-logo-2.png')) }}",
         // Will display country list with corresponding institutions. When `countryFilter` is set to `false`, only list of institutions will be shown.
-        countryFilter: false,
+        countryFilter: true,
```
please explain this change. @turbo124 and I have descided when originaly implementing it to omit the country selection to reduce steps for the user to take.

dshoreman replied:
Both versions have their upsides. Without the country list it's one less click, but the list is highly polluted and scrolling is more difficult because one Bank may have dozens of Institutions in a single country.

With country filtering enabled, it's one more click but the bank lists are much easier to read and scroll through. It's also easy to see at a glance which institutions are supported in your country.

Santander is a good example where country filtering makes the most difference:
![image](https://github.com/user-attachments/assets/7a9e76f5-c7af-4d8f-a4af-38edec96a9a6)
There are 3 *Banco Santander*—in Spain, Brazil and Portugal—but there's no way to know without hovering the link and googling the institution code.

Likewise for *Santander*—one is Germany, the other UK—but without country filtering it's easy to get it wrong.

paulwer replied:
good catch
@turbo124 I would therefore agree to introduce using the country select :)

paulwer commented on app/Helpers/Bank/Nordigen/Nordigen.php:
```diff
@@ -60,22 +60,58 @@ public function getInstitutions()
+     *   accepted: string
+     * } Agreement details
+     */
+    public function getAgreement(string $euaId): array {
+        $eua = $this->client->endUserAgreement->getEndUserAgreement($euaId);
+
+        return $eua;
+    }
```
i have included this method within the PR #10428 and other changes as well, f.ex. to reduce creating of unnecessary new endUserAgreements

please wait until its merged to v5-develop and migrate to the new Nordigen class implementation within your other changes

dshoreman commented:
I have to head out for a few hours but rebase is done. I still need to work @paulwer's EUA method back in which I'll do once I'm back along with other bits.

@turbo124 if you can, please hold off on both PRs for now as things could still change

(trying to join slack but email's not coming through, will try again later)

dshoreman committed:
Attempt to reuse existing agreements

Implements changes from #10410 but using `Arr::first()` rather than a
foreach loop to fix bugs returning invalid or expired agreements.

If an agreement exists with at least the requested `$txDays` then that
is used, otherwise a new one is created with the given parameters. If it
fails, we error out because `createRequisition()` would fail regardless.

Skips accepting EUAs: seems it's done automatically during requisition.

dshoreman committed:
Support setting custom Nordigen access duration

If agreements are already available and no custom `access_days` is
passed, it will simply take the first match regardless of duration.

When a new agreement is needed, it defaults to the max supported.

dshoreman committed:
Show history selection screen instead of guessing

Removes all the guesswork for requisition renewals, instead opting for
the manual approach of simply redirecting to the form and updating its
handling such that it jumps straight to history selection when needed.

When the institution is preloaded (so we skip country/bank selection)
the cloned element is a country and not the bank so we need to replace
athe flag with its logo, but otherwise the code was all moved as-is.

dshoreman commented:
@paulwer Please check over the last 3 commits, especially 5f75b6a.

It merges the changes from #10410 with the code I had; the biggest differences:
* `createRequisition` now only takes `$agreement`, *not* `$txDays`
  * Agreement fetched in controller for better errors (if EUA fails, so would the requisition)
* Replaced agreement loop with a call to `Arr::first()` that filters in a callback function

Second commit 691478d takes care of access days:
* Can be set by adding `&access_days=169`
* If custom access *isn't* set, defaults to...
  * First valid agreement—if there is one—providing it meets `$data['tx_days']` requirement
  * Institution's maximum if an EUA needs to be created

I've also removed my automagical `$txDays` fallback on renewals so it will no longer try to guess the BankIntegration. Instead it takes you back to the form with extra handling to jump to the history selection page.

---

@turbo124 @paulwer **Do we also need UI for the access days, or is it enough to default to max?**

If we need UI, I can see if a slider fits in to avoid another page while also allowing for a default.

Otherwise... this is *just about* ready for review :tada: 

<small>*...except the few lang bits I forgot. CommitPending™*</small>

dshoreman committed:
Add EUA error texts

dshoreman commented:
Lang bits done, think that's everything for now unless I've forgotten something else.

Feel free to throw any questions here or ping on Slack, managed to get it working with an alternative email

paulwer commented:
@dshoreman i dont think we need the ui to select the access duration. Also a hint: import date(after first connect) can also be set when editing a bank_integration to reimport data, so setting it to max kr a default like 30 days and change it if required may be also already be sufficient enought. (my personal opinion)

paulwer commented:
@dshoreman please considder my hints for geting an existing end user agreement. In your implementation you are not reconizing, that the user may is trying to connect another account from the same institution by reusing an already agreed end user agreement. This results in a redirect from the institution by nordigen without the oportunity to login at a different account.

Reusing endUserAgreements is only applicable for not completed flows to reduce orphant endUserAgreements.

Therefore as already recommended, please wait until my PR is merged and then reuse the existing implementation or copy from my branch if you like.

turbo124 commented:
@dshoreman if you are still having troubles connecting to slack, please advise we have had some issues with invites being sent.

paulwer commented:
fyi:
@turbo124 my review is completed and I closed my other PR. all things I wanted to add are now included here as well

turbo124 commented:
@paulwer 

sure thing, @dshoreman is this ready for review?

dshoreman commented on app/Helpers/Bank/Nordigen/Nordigen.php:
```diff
@@ -60,78 +61,128 @@ public function getInstitutions()
-            $endUserAggreements = $this->client->endUserAgreement->getEndUserAgreements();
-        } catch (\Exception $e) { // not able to accept it
-            nlog("Nordigen: Was not able to fetch endUserAgreements. We continue with defaults to setup bank_integration. {$institutionId} {$e->getMessage()} {$e->getCode()}");
+            return Arr::first(
+                $this->client->endUserAgreement->getEndUserAgreements()['results'],
+                function (array $eua) use ($institutionId, $requiredScopes, $accessDays, $txDays): bool {
+                    $expiresAt = $eua['accepted'] ? (new \DateTimeImmutable($eua['accepted']))->add(
+                        new \DateInterval("P{$eua['access_valid_for_days']}D")
+                    ) : false;
+
+                    return $eua['institution_id'] === $institutionId
+                        && $eua['accepted'] === null
+                        && $eua['max_historical_days'] >= $txDays
+                        && $eua['access_valid_for_days'] >= $accessDays
+                        && !array_diff($requiredScopes, $eua['access_scope'] ?? []);
+                },
+                null
+            );
```
@turbo124 There's one change I need to make here, but it depends on yours and @paulwer's thoughts.

The `$expiresAt` line is leftover from my initial implementation before matching it to Paul's second PR. Where it currently checks if `$eua['accepted'] === null`, it would previously use `$expiresAt` to find *any* agreement (as long as it wasn't expired).

**What's the desired behaviour?**
* *Never* try to use an accepted agreement, for creating *or* renewing
  * Current behaviour; just need to remove the `$expiresAt`
* Try to reuse *any* non-expired agreement
  * Original (but I think undesired?) behaviour; simple one-line change
* Hybrid approach that reuses accepted EUAs, but *only* for one flow (renewing?)
  * Maybe better? Not sure, but wouldn't be difficult to implement

If we can clear that up, I'll make the necessary change then it's ready for review/testing

paulwer replied:
I encourage to not use any accepted agreements, as of a user can want to connect different accounts from one institution reusing an existing already agreed agreement may skip the login screen on the institution.

The hybrid aproach would be to much time required to implement / to much complexity to maintain without any benefit for any steakholder.

dshoreman replied:
Sounds good!

dshoreman commented:
> @dshoreman if you are still having troubles connecting to slack, please advise we have had some issues with invites being sent.

I did manage to get in eventually, but I had to use a different address. I want to say it stopped liking aliases (eg `foo+bar@baz.com`), but it took several attempts even with the Gmail account so maybe Slack had a bad day :shrug:

dshoreman committed:
Remove redundant `$expiresAt`

Leftover from a version that could use accepted EUAs.

If this functionality is ever restored, the `$eua['accepted'] === null`
should become `!$eua['accepted'] || $expiresAt > new DateTime('now')`.`

dshoreman commented:
@paulwer Thanks for clarifying, it was the skipping bank part that was confusing me before

@turbo124 Redundant rot removed. Ready for review :)

turbo124 commented:
@dshoreman looks good, thanks!
